### PR TITLE
fix: allow dynamic item count

### PR DESCRIPTION
**Pull request by tanguyantoine**
# What

Add a `useEffect` hook that populate items refs.

# Why

It allows dynamic item count

use case

```tsx
const MyCompontent = () => {
  const [items, setItems] = useState(['1', '2'])
  const { buttonProps, itemProps, isOpen, setIsOpen } = useDropdownMenu(items.length)
  setTimeout(() => {
    setItems([...items, '3'])
  }, 3000)

  return (
    <ul>
      {items.map((item, index) => (
        <li key={item} {...itemProps[index]}>
          <a>{item}</a>
        </li>
      ))}
    </ul>
  )
}
```

### CODE REVIEW

**Comment by tanguyantoine** [comment]
interesting, not adding code increases the coverage 🤣

File: src/use-dropdown-menu.ts [hunk 1]
```diff
@@ -20,7 +20,8 @@ export default function useDropdownMenu(itemCount: number) {
 
 	// Create refs
 	const buttonRef = useRef<HTMLButtonElement>(null);
-	const itemRefs = useRef([...Array(itemCount)].map(() => createRef<HTMLAnchorElement>()));
+
```
Comment (WesCossick) [inline src/use-dropdown-menu.ts:23]:
```suggestion
```

File: src/use-dropdown-menu.ts [hunk 2]
```diff
@@ -32,6 +33,11 @@ export default function useDropdownMenu(itemCount: number) {
 		itemRefs.current[itemIndex].current?.focus();
 	};
 
+	// Initialize refs an update them on prop changes
```
Comment (WesCossick) [inline src/use-dropdown-menu.ts:36]:
```suggestion
	// Initialize refs and update them when the item count changes
```

**Review by WesCossick — CHANGES_REQUESTED** [review]
Thanks @tanguyantoine for submitting this! I made a couple of minor suggestions. Most importantly, though, can you add a unit test that would fail with our existing code, but is fixed by your changes?

This will help us avoid reintroducing this bug in the future if we ever refactor the code.

**Comment by corymharper** [comment]
Good contribution @tanguyantoine!

I do have a concern about the current approach though.

Consider the following scenario:

```js
    const [items, setItems] = useState(['1', '2'..., '10,000'])
    const { buttonProps, itemProps, isOpen, setIsOpen } = useDropdownMenu(items.length)
```

In the above scenario, the hook would originally create an array of 10,000 refs, and that array would never change. The same thing happens on the first render in your current approach, but if they then add one item to the array, it scraps the old array, and creates 10,001 new refs to fill the array. 

This is an extreme example, but I think it showcases my concern, is instantiating a whole new array of refs the most optimal method for increasing or decreasing the number of refs in the array? The additional steps needed to implement a different solution may undercut any performance gains we would get from a different implementation, but I think it's worth having a discussion about.

**Comment by tanguyantoine** [comment]
Would you prefer `Array.from(Array(itemCount))` ?

**Comment by tanguyantoine** [comment]
Let's be realistic, why would you put 10 000 items in a dropdown? 
Let's assume you dot it, we would regenerate an array of ref if the itemCount value changes. At the moment it does not work, it is expected not to be able to change the number of items?

**Comment by corymharper** [comment]
I agree with you that it's not a realistic use case, so we probably don't need to consider it. If you could review the couple of changes that @WesCossick requested, on line 23 and 36, and add a test for the change you made, simply one that verifies the ref array length is changing when items are removed or added as expected should suffice, I think this PR looks good to be merged in!

**Comment by tanguyantoine** [comment]
I'll do it as soon as I get bandwidth for it

**Pull request closed without merging**